File: backend/app/schemas/template.py
import uuid

from pydantic import BaseModel, ConfigDict

from app.models.enums import TemplateScope


class TemplateBase(BaseModel):
    name: str
    description: str | None = None
    scope: TemplateScope
    is_public: bool = False


class TemplateCreate(TemplateBase):
    workspace_id: uuid.UUID | None = None
    snapshot: dict


class TemplateRead(TemplateBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workspace_id: uuid.UUID | None
    snapshot: dict
    created_by: uuid.UUID


class TemplateCaptureBoardRequest(BaseModel):
    name: str
    description: str | None = None
    workspace_id: uuid.UUID
    is_public: bool = False


class TemplateCaptureCardRequest(BaseModel):
    name: str
    description: str | None = None
    workspace_id: uuid.UUID
    is_public: bool = False


class TemplateApplyRequest(BaseModel):
    workspace_id: uuid.UUID | None = None
    list_id: uuid.UUID | None = None
    name: str | None = None


class TemplateApplyResult(BaseModel):
    board_id: uuid.UUID | None = None
    card_id: uuid.UUID | None = None

File: frontend/src/types/template.ts
export type TemplateScope = "board" | "card";

export interface Template {
  id: string;
  name: string;
  description: string | null;
  scope: TemplateScope;
  is_public: boolean;
  workspace_id: string | null;
  snapshot: Record<string, unknown>;
  created_by: string;
}

export interface TemplateCaptureBoardRequest {
  name: string;
  description?: string | null;
  workspace_id: string;
  is_public?: boolean;
}

export interface TemplateCaptureCardRequest {
  name: string;
  description?: string | null;
  workspace_id: string;
  is_public?: boolean;
}

export interface TemplateApplyRequest {
  workspace_id?: string | null;
  list_id?: string | null;
  name?: string | null;
}

export interface TemplateApplyResult {
  board_id: string | null;
  card_id: string | null;
}
